src/boot/efi/devicetree.c

```diff
@@ -104,6 +113,90 @@ EFI_STATUS devicetree_install(struct devicetree_state *state, EFI_FILE *root_dir
+        assert(dtb);
+        const struct fdt_header *dt_header = dtb;
+        if (!IS_ALIGNED64(dt_header) || be32toh(dt_header->Magic) != UINT32_C(0xd00dfeed))
+                return NULL;
+
+        uint32_t dt_size = be32toh(dt_header->TotalSize);
+        uint32_t struct_off = be32toh(dt_header->OffDTStruct);
+        uint32_t struct_size = be32toh(dt_header->SizeDTStruct);
+        uint32_t strings_off = be32toh(dt_header->OffDTStrings);
+        uint32_t strings_size = be32toh(dt_header->SizeDTStrings);
+        uint32_t tmp;
+
+        if (struct_off % sizeof(uint32_t) != 0 || struct_size % sizeof(uint32_t) != 0 ||
+            check_add_overflow(strings_off, strings_size, &tmp) || tmp > dt_size ||
+            check_add_overflow(struct_off, struct_size, &tmp) || tmp > strings_off)
+                return NULL;
+
+        const uint32_t *cursor = (const uint32_t *) ((uint8_t *) dt_header + struct_off);
+        const char *strings_block = (const char *) ((uint8_t *) dt_header + strings_off);
+
+        size_t size_words = struct_size / sizeof(uint32_t);
+        if (check_sub_overflow(size_words, 3, &tmp))
+                return NULL;
+
+        for (size_t i = 0; i < tmp; i++) {
+                switch (be32toh(cursor[i])) {
+                case FDT_BEGIN_NODE:
+                        if (cursor[++i] != 0)
+                                return NULL;
+                        break;
+                case FDT_NOP:
+                        break;
+                case FDT_PROP:
+                        size_t len = be32toh(cursor[++i]);
+                        size_t name_off = be32toh(cursor[++i]);
+                        size_t len_words = DIV_ROUND_UP(len, sizeof(uint32_t));
+                        size_t s;
+
+                        if (!check_add_overflow(name_off, strlen8("compatible"), &s) &&
+                            s < strings_size && streq8(strings_block + name_off, "compatible")) {
+                                const char *c = (const char *) &cursor[++i];
+                                if (len == 0 || i + len_words > size_words || c[len - 1] != '\0')
+                                        c = NULL;
+
+                                return c;
+                        }
+                        i += len_words;
+                        break;
+                default:
+                        return NULL;
+
+                }
+        }
+
+        return NULL;
+}
+
+EFI_STATUS devicetree_match(const void *dtb_buffer, size_t dtb_length) {
+        assert(dtb_buffer);
+        const struct fdt_header *dt_header = (const struct fdt_header *)dtb_buffer;
+
+        const void *fw_dtb = find_configuration_table(MAKE_GUID_PTR(EFI_DTB_TABLE));
+        if (!fw_dtb)
+                return EFI_UNSUPPORTED;
+
+        const char *fw_compat = devicetree_get_compatible(fw_dtb);
+        if (dtb_length < sizeof(struct fdt_header) ||
+            dtb_length < be32toh(dt_header->TotalSize) ||
```

Comment:
so this is weird. you call devicetree_get_compatible() here which assumes the structure is complete and derefs all kinds of fields in there. And then after all that ran you check the size of it. That's too late! you already accessed the possibly invalid memory.

Generally the rule is: check sizes *before* you deref pointers, not afterwards.

I mean, I appreciate that if the DT comes frm the config table you don't have any size to validate, and that's entirely fine, but if it comes from the PE section you do have a size and you should take it into consideration before processing the table.

hence, I'd propose that devicetree_get_compatible() gets the size passed, and if you don't know it you pass SIZE_MAX there as indicator for "don't know".

And inside of devicetree_get_compatible() you always check the size *before* you deref the structure, and if needed special case the SIZE_MAX case any bypass that specific check.

Reply:
Here `dtb_length` refers to the PE-provided DT blob, so the checks refer to the following function call `const char *compat = devicetree_get_compatible(dtb_buffer)`. I put them here to try and make the checks more compact with the `!fw_compat` check since they return the same value. I think it makes some sense that the size check happens in this function since we are already passed the length here and it makes `devicetree_get_compatible()` PE-independent, but I can move the check there if it makes it more readable, or split the size checks from `!fw_compat`.

Reply:
Yeah, please split the checks to reject invalid `dtb_buffer` immediately at the top of the function.